distribution/lib/Standard/Google_Api/0.0.0-dev/src/Main.enso

```diff
@@ -58,3 +67,36 @@ initialize secret_file =
     http_transport = GoogleNetHttpTransport.newTrustedTransport
     json_factory = GsonFactory.getDefaultInstance
     Google_Api_Client.Value credential json_factory http_transport
+
+
+## PLACEHOLDER, performs google analytics call
+
+   Arguments:
+   - property_id
+   - start_date
+   - end_date
+run_google_report : Text -> Date -> Date -> Table
+run_google_report property_id:Text start_date:Date=(Date.new 2020 3 31) end_date:Date=Date.today =
+    analytics_data = BetaAnalyticsDataClient.create
+    request = RunReportRequest.newBuilder
+        . setProperty ("properties/"+property_id)
+        . addDimensions (Dimension.newBuilder.setName "city")
+        . addMetrics (Metric.newBuilder.setName "activeUsers")
+        . addDateRanges (DateRange.newBuilder.setStartDate start_date.to_text . setEndDate end_date.to_text)
+        . build
+    response = analytics_data.runReport request
+
+    dimension_count = response.getDimensionHeadersCount
+    dimensions = 0.up_to dimension_count . map i-> response.getDimensionHeaders i . getName
+    metric_count = response.getMetricHeadersCount
+    metrics = 0.up_to metric_count . map i-> response.getMetricHeaders i . getName
+    headers = dimensions + metrics
+
+    row_count = response.getRowCount
+    row_proxy = Array_Proxy.new row_count i->
+        row = response.getRows i
+        dimension_values = 0.up_to dimension_count . map i-> row.getDimensionValues i . getValue
+        metric_values = 0.up_to metric_count . map i-> row.getMetricValues i . getValue
+        dimension_values + metric_values
```

Comment:
I'm not sure `Array_Proxy` is the right choice here.

There is quite a lot of computation going on in here and it will be repeated every time a row from the returned vector is requested.

Doesn't it make more sense to compute it just once and save it in the vector, allowing the `response` object to be garbage collected?

Reply:
IMO Array_Proxy makes sense if the values can be computed relatively cheaply and the collection is very big, so we can save memory.

Here my intuition suggests that the `map`s inside of the row computation are not that cheap to compute and it may actually make sense to save them in a vector instead of recomputing on each access.

I don't think we are saving memory either because the Array_Proxy holding a reference to the `response` makes it impossible to garbage collect, so roughly similar amount of memory is probably used if we just mapped.

It would make sense I guess if we only fetch a few rows from the response and not most of them. Is that maybe the case (I'm not very familiar with the end use-case of this).

Reply:
It's read through once and loaded into a Table, but yes we could return an Array_Proxy for each row instead instead of building vector explicitly. One we can do when we mature the API later.

Reply:
Will add to the backlog as a piece of the follow-on request